File: app/src/main/java/statusbar/finder/sql/OriginManager.java
package statusbar.finder.sql;

import android.database.sqlite.SQLiteConstraintException;
import statusbar.finder.DatabaseHelper;
import statusbar.finder.Origin;
import statusbar.finder.OriginQueries;
import statusbar.finder.provider.ILrcProvider;

/**
 * LyricGetterExt - statusbar.finder.sql
 *
 * @author VictorModi
 * @description TODO: coming soon.
 * @email victormodi@outlook.com
 * @date 2025/1/20 下午5:15
 */
public class OriginManager {
    private static OriginQueries queries = null;

    public static Long insertOrGetMediaInfoId(ILrcProvider.MediaInfo mediaInfo, String packageName) {
        if (queries == null) {queries = DatabaseHelper.getDatabase().getOriginQueries();}
        Long result = getMediaInfoId(mediaInfo, packageName);
        if (result != null) {return result;}
        try {
            queries.insertMediaInfo(
                    mediaInfo.getTitle(),
                    mediaInfo.getArtist(),
                    mediaInfo.getAlbum(),
                    mediaInfo.getDuration(),
                    packageName
            );
            result = queries.getLastInsertId().executeAsOne();
            return result;
        } catch (SQLiteConstraintException e) {
            e.fillInStackTrace();
            return (long) -1;
        }
    }

    private static Long getMediaInfoId(ILrcProvider.MediaInfo mediaInfo, String packageName) {
        if (queries == null) {queries = DatabaseHelper.getDatabase().getOriginQueries();}
        return queries.getMediaInfoId(
                mediaInfo.getTitle(),
                mediaInfo.getArtist(),
                mediaInfo.getAlbum(),
                packageName
        ).executeAsOneOrNull();
    }

    public static ILrcProvider.MediaInfo getMediaInfoById(Long id) {
        if (queries == null) {queries = DatabaseHelper.getDatabase().getOriginQueries();}
        Origin data = queries.getInfoById(id).executeAsOneOrNull();
        if (data == null) {return null;}
        return new ILrcProvider.MediaInfo(data);
    }
}

File: app/src/main/java/statusbar/finder/sql/OriginManager.kt
package statusbar.finder.sql

import android.database.sqlite.SQLiteConstraintException
import statusbar.finder.DatabaseHelper
import statusbar.finder.OriginQueries
import statusbar.finder.data.MediaInfo

/**
 * LyricGetterExt - statusbar.finder.sql
 * @description TODO: coming soon.
 * @author VictorModi
 * @email victormodi@outlook.com
 * @date 2025/2/9 13:52
 */
object OriginManager {
    private val queries: OriginQueries by lazy {
        DatabaseHelper.database?.originQueries
            ?: throw IllegalStateException("Database not initialized")
    }

    @JvmStatic
    fun insertOrGetMediaInfoId(mediaInfo: MediaInfo, packageName: String): Long {
        getOriginId(mediaInfo, packageName)?.let { return it }
        return try {
            queries.insertMediaInfo(
                mediaInfo.title,
                mediaInfo.artist,
                mediaInfo.album,
                mediaInfo.duration,
                packageName
            )
            queries.getLastInsertId().executeAsOne()
        } catch (e: SQLiteConstraintException) {
            e.printStackTrace()
            -1L
        }
    }

    fun getOriginId(mediaInfo: MediaInfo, packageName: String): Long? {
        return queries.getMediaInfoId(
            mediaInfo.title,
            mediaInfo.artist,
            mediaInfo.album,
            packageName
        ).executeAsOneOrNull()
    }

    @JvmStatic
    fun getMediaInfoById(id: Long): MediaInfo? {
        return queries.getInfoById(id).executeAsOneOrNull()?.let { MediaInfo(it) }
    }
}
